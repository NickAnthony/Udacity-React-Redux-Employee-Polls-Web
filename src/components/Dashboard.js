import { connect } from "react-redux";
import QuestionsTabLayout from "./QuestionsTabLayout";

const Dashboard = ({ questionIds, questions, authedUser }) => {
  const answeredQuestionIds = questionIds.filter(
    (questionId) =>
      !questions[questionId].optionOne.votes.includes(authedUser) &&
      !questions[questionId].optionTwo.votes.includes(authedUser)
  );

  const unansweredQuestionIds = questionIds.filter(
    (questionId) =>
      questions[questionId].optionOne.votes.includes(authedUser) ||
      questions[questionId].optionTwo.votes.includes(authedUser)
  );

  return (
    <div className="container dashboard-container">
      <QuestionsTabLayout
        tabs={["Unanswered Questions", "Answered Questions"]}
        questionsPerTab={{
          0: answeredQuestionIds,
          1: unansweredQuestionIds,
        }}
      />
    </div>
  );
};

const mapStateToProps = ({ questions, authedUser }) => {
  return {
    questionIds: Object.keys(questions),
    questions,
    authedUser,
  };
};

export default connect(mapStateToProps)(Dashboard);
