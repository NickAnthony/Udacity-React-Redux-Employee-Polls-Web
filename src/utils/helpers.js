import { useLocation, useNavigate, useParams } from "react-router-dom";

// Copied directly from https://reactrouter.com/en/main/start/faq
export function withRouter(Component) {
  function ComponentWithRouterProp(props) {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  }

  return ComponentWithRouterProp;
}

export function formatTimestamp(timestamp) {
  const date = new Intl.DateTimeFormat("en-US", {
    year: "numeric",
    month: "numeric",
    day: "numeric",
  }).format(timestamp);
  const time = new Intl.DateTimeFormat("en-US", {
    hour: "numeric",
    minute: "2-digit",
  }).format(timestamp);
  return `${time} | ${date}`;
}
