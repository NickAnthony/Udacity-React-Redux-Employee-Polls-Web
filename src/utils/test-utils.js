import { configureStore } from "@reduxjs/toolkit";
import reducers from "../reducers";
import middleware from "../middleware";
import { users, questions } from "../utils/_DATA";
import CreateQuestion from "../components/CreateQuestion";
import CreateUser from "../components/CreateUser";
import Dashboard from "../components/Dashboard";
import Leaderboard from "../components/Leaderboard";
import Login from "../components/Login";
import Question from "../components/Question";
import UserProfile from "../components/UserProfile";
import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";

export function createTestStore(initialState) {
  const store = configureStore({
    reducer: reducers,
    middleware: middleware,
    preloadedState: initialState,
  });
  return store;
}

export function createTestStoreWithInitialData() {
  return createTestStore({
    users,
    questions,
    authedUser: "nickanthony",
  });
}

export function createTestAppWithStore(initialRoute) {
  return (
    <Provider store={createTestStoreWithInitialData()}>
      <MemoryRouter initialEntries={[initialRoute]}>
        <Routes>
          <Route path="/poll/:id" element={<Question />} />
        </Routes>
      </MemoryRouter>
    </Provider>
  );
}
