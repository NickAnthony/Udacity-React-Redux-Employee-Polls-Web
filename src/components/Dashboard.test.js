import { render } from "@testing-library/react";
import { createTestAppWithStoreAndRouter } from "../utils/test-utils";
import Dashboard from "./Dashboard";

describe("Dashboard", () => {
  it("will display the vote counts by clicking either of the vote buttons.", () => {
    var component = render(createTestAppWithStoreAndRouter("/"));
    expect(component).toMatchSnapshot();
  });
});
