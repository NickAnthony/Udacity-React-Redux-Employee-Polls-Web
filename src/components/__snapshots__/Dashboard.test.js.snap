// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard will display the vote counts by clicking either of the vote buttons. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container dashboard-container"
      >
        <div
          class="container dashboard-container"
        >
          <div
            class="container-row tab-container"
          >
            <div
              class="tab-button tab-button-selected"
              tabindex="0"
            >
              Unanswered Questions
            </div>
            <div
              class="tab-button "
              tabindex="1"
            >
              Answered Questions
            </div>
          </div>
          <div
            class="container-row-start question-layout-container negative-top-margin"
          >
            <div
              class="question-card-container"
            >
              <div
                class="question-card-name"
              >
                sarahedo
              </div>
              <div
                class="question-card-timestamp"
              >
                10:21 PM | 6/28/2016
              </div>
              <button
                class="question-card-show-btn"
              >
                Show
              </button>
            </div>
            <div
              class="question-card-container"
            >
              <div
                class="question-card-name"
              >
                mtsamis
              </div>
              <div
                class="question-card-timestamp"
              >
                3:02 AM | 7/14/2016
              </div>
              <button
                class="question-card-show-btn"
              >
                Show
              </button>
            </div>
            <div
              class="question-card-container"
            >
              <div
                class="question-card-name"
              >
                sarahedo
              </div>
              <div
                class="question-card-timestamp"
              >
                5:22 PM | 3/3/2017
              </div>
              <button
                class="question-card-show-btn"
              >
                Show
              </button>
            </div>
            <div
              class="question-card-container"
            >
              <div
                class="question-card-name"
              >
                tylermcginnis
              </div>
              <div
                class="question-card-timestamp"
              >
                6:42 AM | 12/24/2016
              </div>
              <button
                class="question-card-show-btn"
              >
                Show
              </button>
            </div>
            <div
              class="question-card-container"
            >
              <div
                class="question-card-name"
              >
                tylermcginnis
              </div>
              <div
                class="question-card-timestamp"
              >
                8:09 AM | 3/15/2017
              </div>
              <button
                class="question-card-show-btn"
              >
                Show
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container dashboard-container"
    >
      <div
        class="container dashboard-container"
      >
        <div
          class="container-row tab-container"
        >
          <div
            class="tab-button tab-button-selected"
            tabindex="0"
          >
            Unanswered Questions
          </div>
          <div
            class="tab-button "
            tabindex="1"
          >
            Answered Questions
          </div>
        </div>
        <div
          class="container-row-start question-layout-container negative-top-margin"
        >
          <div
            class="question-card-container"
          >
            <div
              class="question-card-name"
            >
              sarahedo
            </div>
            <div
              class="question-card-timestamp"
            >
              10:21 PM | 6/28/2016
            </div>
            <button
              class="question-card-show-btn"
            >
              Show
            </button>
          </div>
          <div
            class="question-card-container"
          >
            <div
              class="question-card-name"
            >
              mtsamis
            </div>
            <div
              class="question-card-timestamp"
            >
              3:02 AM | 7/14/2016
            </div>
            <button
              class="question-card-show-btn"
            >
              Show
            </button>
          </div>
          <div
            class="question-card-container"
          >
            <div
              class="question-card-name"
            >
              sarahedo
            </div>
            <div
              class="question-card-timestamp"
            >
              5:22 PM | 3/3/2017
            </div>
            <button
              class="question-card-show-btn"
            >
              Show
            </button>
          </div>
          <div
            class="question-card-container"
          >
            <div
              class="question-card-name"
            >
              tylermcginnis
            </div>
            <div
              class="question-card-timestamp"
            >
              6:42 AM | 12/24/2016
            </div>
            <button
              class="question-card-show-btn"
            >
              Show
            </button>
          </div>
          <div
            class="question-card-container"
          >
            <div
              class="question-card-name"
            >
              tylermcginnis
            </div>
            <div
              class="question-card-timestamp"
            >
              8:09 AM | 3/15/2017
            </div>
            <button
              class="question-card-show-btn"
            >
              Show
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
